##################################################################
##  ______     ______     __  __                                ##
## /\  __ \   /\  ___\   /\ \/\ \                               ##
## \ \  __ \  \ \___  \  \ \ \_\ \                              ##
##  \ \_\ \_\  \/\_____\  \ \_____\                             ##
##   \/_/\/_/   \/_____/   \/_____/                             ##
## @Developed by: Plaisier Lab                                  ##
##   (https://plaisierlab.engineering.asu.edu/)                 ##
##   Arizona State University                                   ##
##   242 ISTB1, 550 E Orange St                                 ##
##   Tempe, AZ  85281                                           ##
## @github: https://github.com/A-Zehnder/scSignalMap            ##
## @Author: Jillian Woodley, Samantha O'Connor, Chris Plaisier  ##
## @License:  GNU GPLv3                                         ##
##                                                              ##
## If this program is used in your analysis please              ##
## mention who built it. Thanks. :-)                            ##
##################################################################
#' Capturing signaling pathways using scRNA-seq data
#'
#' Function to capture signaling pathways using scRNA-seq data.
#'
#'
#' @param seurat_obj: a seurat object must be supplied to classify, no default
#' @param group_by: a meta_data column that will be used to split the cells, no default
#' @param avg_log2FC_gte: 
#' @param p_val_adj_lte: 
#' @param species: from which species did the samples originate, either 'human' or 'mouse', defaults to 'human'
#' @param gene_id: what type of gene ID is used, either 'ensembl' or 'symbol', defaults to 'ensembl'
#' @return data.frame with putative interactions
#' @export
MapInteractions = function(seurat_obj, group_by, avg_log2FC_gte = 0.25, p_val_adj_lte = 0.05, min_pct = 0.1, species='human', gene_id='ensembl') {
    cat('Running scSignalMap:\n')

    
    ## Step 1. Load up the data
    cat('  Loading data...\n')
    # Load MultiNicheNet ligand receptor interactions
    lr_network = read.csv(system.file('extdata', 'lr_network.csv', package='scSignalMap'), header=TRUE)
    lr_pairs = na.omit(lr_network[,c(paste('ligand',species,gene_id,sep='_'),paste('receptor',species,gene_id,sep='_'))])
    #cat(1)
    if(gene_id!='symbol') {
        lig_convert = lr_network[,c(paste('ligand',species,'symbol',sep='_'))]
        names(lig_convert) = lr_network[,c(paste('ligand',species,gene_id,sep='_'))]
        rec_convert = lr_network[,c(paste('receptor',species,'symbol',sep='_'))]
        names(rec_convert) = lr_network[,c(paste('receptor',species,gene_id,sep='_'))]
    }

    # Load up secreted ligands
    secreted = read.csv(system.file('extdata', 'secreted.csv', package='scSignalMap'), header=TRUE)
    colnames(secreted)[1] = 'human_ensembl'
    secreted_ligands = na.omit(secreted[,paste(species,gene_id,sep='_')])


    ## Step 2. Identify marker genes
    cat('  Identifying marker genes...\n')
    
    putative_markers = FindAllMarkers(seurat_obj, group.by=group_by, min.pct=min_pct, verbose=T)
    markers = list()
    for(clust1 in as.character(sort(unique(seurat_obj@meta.data[,group_by])))) {
        markers[[clust1]] = putative_markers %>% dplyr::filter(cluster==clust1 & avg_log2FC>=avg_log2FC_gte & p_val_adj<=p_val_adj_lte) %>% dplyr::pull(name='gene')
        #cat(paste0('    ',cluster1,': ',length(markers[[cluster1]]),'\n'))
    }


    ## Step 3. Precompute
    cat('  Precomputing...\n')
    
    # Get a vector of marker genes   
    allgenes = rownames(seurat_obj@assays$RNA)[which(rowSums(as.matrix(seurat_obj@assays$RNA$counts))>0)]

    # Subset ligand_receptor list to those in allgenes
    lr_pairs = lr_pairs[lr_pairs[,1] %fin% allgenes,]
    lr_pairs = lr_pairs[lr_pairs[,2] %fin% allgenes,]

    # Split up the seurat object by groups
    seurat_obj_split = SplitObject(seurat_obj, split.by=group_by)

    # Build a data.table with statistics
    all_dt = rbindlist(lapply(as.character(sort(unique(seurat_obj@meta.data[,group_by]))), function(clust1) {
                 mat = as.matrix(seurat_obj_split[[clust1]]@assays$RNA@layers$counts)
                 genes = rownames(seurat_obj_split[[clust1]]@assays$RNA)
                 n_cells = ncol(mat)

                 data.table(
                     clust1    = clust1,
                     gene      = genes,
                     counts    = rowSums(mat),
                     perc_gt_0  = rowSums(mat > 0)  / n_cells,
                     perc_gte_3 = rowSums(mat >= 3) / n_cells,
                     perc_gte_10= rowSums(mat >= 10)/ n_cells,
                     avg_exp   = rowMeans(mat)
                 )
             }), use.names = TRUE)


    ## Step 4. Integrate ligand receptor pair with expression data
    
    cat('  L-R and S-R pairs...\n')
    
    
    # All sender/receiver combinations
    clusts = as.character(sort(unique(seurat_obj@meta.data[[group_by]])))
    clust_dt = data.table::CJ(Sender = clusts, Receiver = clusts)  # Cartesian product

    # Ligand-Receptor pairs table
    if (gene_id == "symbol") {
      lr_dt = data.table::data.table(
        Ligand   = lr_pairs[, 1],
        Receptor = lr_pairs[, 2]
      )
    } else {
      lr_dt = data.table::data.table(
        Ligand          = lr_pairs[, 1],
        Ligand_Symbol   = lig_convert[lr_pairs[, 1]],
        Receptor        = lr_pairs[, 2],
        Receptor_Symbol = rec_convert[lr_pairs[, 2]]
      )
    }

    # Cartesian product of LR pairs x cluster pairs
    lr_dt[, `:=`(dummy, 1)]
    clust_dt[, `:=`(dummy, 1)]

    pairs_data = lr_dt[clust_dt, on = "dummy", allow.cartesian = TRUE][, `:=`(dummy, NULL)]
    #pairs_data = lr_dt[, cbind(.SD, clust_dt), by = seq_len(nrow(lr_dt))][, `:=`(seq_len, NULL)]

    cat(paste0('    Rows = ',nrow(pairs_data),'\n'))

    cat('  Integrating data...\n')
    steps = 13
    pb = utils::txtProgressBar(min = 0, max = steps, style = 3)
    pairs_data[, `:=`(Ligand_Counts, all_dt[pairs_data, on = .(clust1=Sender, gene=Ligand), counts])]
    utils::setTxtProgressBar(pb, 1)
    pairs_data[, `:=`(Ligand_gte_3, all_dt[pairs_data, on = .(clust1=Sender, gene=Ligand), perc_gte_3])]
    utils::setTxtProgressBar(pb, 2)
    pairs_data[, `:=`(Ligand_gte_10, all_dt[pairs_data, on = .(clust1=Sender, gene=Ligand), perc_gte_10])]
    utils::setTxtProgressBar(pb, 3)
    pairs_data[, `:=`(Ligand_Cells_Exp, all_dt[pairs_data, on = .(clust1=Sender, gene=Ligand), perc_gt_0])]
    utils::setTxtProgressBar(pb, 4)
    pairs_data[, `:=`(Ligand_Avg_Exp, all_dt[pairs_data, on = .(clust1=Sender, gene=Ligand), avg_exp])]
    utils::setTxtProgressBar(pb, 5)
    pairs_data[, `:=`(Ligand_Cluster_Marker, mapply(function(sender, ligand) { ligand %in% markers[[sender]] }, pairs_data$Sender, pairs_data$Ligand))]
    utils::setTxtProgressBar(pb, 6)
    pairs_data[, `:=`(Ligand_secreted, sapply(pairs_data[['Ligand']], function(x) { x %fin% secreted_ligands }))]
    utils::setTxtProgressBar(pb, 7)
    pairs_data[, ':='(Receptor_Counts, all_dt[pairs_data, on = .(clust1=Receiver, gene=Receptor), counts])]
    utils::setTxtProgressBar(pb, 8)
    pairs_data[,`:=`(Receptor_gte_3, all_dt[pairs_data, on = .(clust1=Receiver, gene=Receptor), 'perc_gte_3'])]
    utils::setTxtProgressBar(pb, 9)
    pairs_data[,`:=`(Receptor_gte_10, all_dt[pairs_data, on = .(clust1=Receiver, gene=Receptor), 'perc_gte_10'])]
    utils::setTxtProgressBar(pb, 10)
    pairs_data[,`:=`(Receptor_Cells_Exp, all_dt[pairs_data, on = .(clust1=Receiver, gene=Receptor), perc_gt_0])]
    utils::setTxtProgressBar(pb, 11)
    pairs_data[, `:=`(Receptor_Avg_Exp, all_dt[pairs_data, on = .(clust1=Receiver, gene=Receptor), avg_exp])]
    utils::setTxtProgressBar(pb, 12)
    pairs_data[, `:=`(Receptor_Cluster_Marker, mapply(function(receiver, receptor) { receptor %in% markers[[receiver]] }, pairs_data$Receiver, pairs_data$Receptor))]
    utils::setTxtProgressBar(pb, 13)
    close(pb)
    cat('Done.\n')

    return(pairs_data)
}


##' Perform Differential Expression for a Desired Celltype
#'
#' This function identifies differentially expressed (DE) genes between two conditions 
#' within a specific celltype of interest. It uses Seurat's `FindMarkers` function with 
#' customizable log2FC and adjusted p-value cutoffs. The results are filtered and 
#' annotated with gene symbols from Ensembl IDs. 
#'
#' @param seurat_obj: seurat object containing scRNA data
#' @param prep_SCT: specify whether the PrepSCTFindMarker function needs to be applied, defaults to FALSE
#' @param cond_column: meta.data column name for condition information 
#' @param cond_name1: condition column value that compared to cond_name2 for differential expression analysis, must be specified
#' @param cond_name2: condition column value that compared to cond_name1 for differential expression analysis, must be specified
#' @param celltype_column: meta.data column name for celltype information
#' @param celltype_name: celltype column value of interest, must be specified 
#' @param FC_cutoff: desired cutoff for log2FC values using >= the absolute value, default is 0.3
#' @param adj_p_val_cutoff: desired cutoff for the adjusted p value, default is 0.05
#' @return A data frame of DE genes between conditions for specified cell type, including Ensembl ID, gene symbol, log2FC, and adjusted p-value.
#' @export
find_markers_btwn_cond_for_celltype = function(seurat_obj = NULL, prep_SCT = FALSE, cond_column = NULL, cond_name1 = NULL, cond_name2 = NULL, celltype_column = NULL, celltype_name = NULL, FC_cutoff = 0.3, adj_p_val_cutoff = 0.05, ensdb = 'EnsDb.Hsapiens.v86') {

    #message("Preparing to run FindMarkers...")
    #if(prep_SCT==TRUE) {
    #    seurat_obj = PrepSCTFindMarkers(seurat_obj)
    #}

    message("Subsetting and setting identities...")
    cells_to_keep = rownames(seurat_obj@meta.data)[seurat_obj@meta.data[,celltype_column] == celltype_name]
    subset_cells = subset(seurat_obj, cells = cells_to_keep)
    Idents(subset_cells) = subset_cells@meta.data[,cond_column]

    message("Running FindMarkers...")
    subset_cells = PrepSCTFindMarkers(subset_cells)
    de_cells = FindMarkers(subset_cells, ident.1 = cond_name1, ident.2 = cond_name2, recorrect_umi = FALSE)

    message("Filtering DE genes by log2FC and adjusted p-value...")
    de_cond_celltype = de_cells %>%
                       dplyr::filter((avg_log2FC >= FC_cutoff | avg_log2FC <= -FC_cutoff) & p_val_adj <= adj_p_val_cutoff)

    message("Adding gene symbols...")
    ensembl_ids = rownames(de_cond_celltype)
    
    gene_symbols = AnnotationDbi::select(eval(parse(text=ensdb)), keys = ensembl_ids, keytype = "GENEID", columns = c("SYMBOL"), multiVals='asNA')
    rownames(gene_symbols) = gene_symbols[,'GENEID']

    de_cond_celltype = data.frame(ensembl_id = ensembl_ids,
                                  gene_symbol = gene_symbols[ensembl_ids,'SYMBOL'],
                                  de_cond_celltype,
                                  stringsAsFactors = FALSE)

    de_cond_celltype = de_cond_celltype[, c("ensembl_id", "gene_symbol", setdiff(colnames(de_cond_celltype), c("ensembl_id", "gene_symbol")))]

    return(de_cond_celltype)
}


##' Identify upregulated receptors
#'
#' This function identifies upregualted receptors from a given DE gene table using a chosen log2FC cutoff
#'
#' @param de_condition_filtered: differentially expressed genes output from find_markers_btwn_cond_for_celltype function
#' @param FC_cutoff: desired cutoff for log2FC values using >= the absolute value, default is 0.3
#' @return A dataframe with identified DE genes and their log2FC from previously chosen condition
#' @export
find_upreg_receptors = function(de_condition_filtered= NULL, FC_cutoff = 0.3, species = 'human') {

    message("Loading ligand-receptor information")
    # Load MultiNicheNet ligand receptor interactions
    lr_network = read.csv(system.file('extdata', 'lr_network.csv', package='scSignalMap'), header=TRUE)
    receptor_ensembl = lr_network[,paste('receptor',species,'ensembl',sep='_')]
    receptor_symbol = lr_network[,paste('receptor',species,'symbol',sep='_')]
    receptor_genes = unique(na.omit(receptor_ensembl))
    ensembl_to_symbol = setNames(receptor_symbol, receptor_ensembl)

    message("Filter for upregulated receptors")
    upreg_receptors = de_condition_filtered %>%
                      dplyr::filter(ensembl_id %in% receptor_genes & avg_log2FC >= FC_cutoff)
    upreg_receptors$gene_symbol = ensembl_to_symbol[upreg_receptors$ensembl_id]

    upreg_receptors = upreg_receptors[, c("gene_symbol", setdiff(names(upreg_receptors), "gene_symbol"))]  

    return(upreg_receptors)
}


##' Filter identified ligand-receptor interactions
#'
#' This function filters interactions from a scSignalMap output based on celltype and ligand secretion
#'
#' @param interactions: ligand-receptor interactions dataframe from MapInteractions function of scSignalMap
#' @param sender_celltypes: celltype of cell(s) that are "senders" in the cell-to-cell communication interaction, ex. "<celltype>" or c("celltype1", "celltype2")
#' @param receiver_celltypes: celltype of cell(s) that are "receivers" in the cell-to-cell communication interaction, ex. "<celltype>" or c("celltype1", "celltype2")  
#' @param secreted_lig: logical statement if the ligand in the interaction is secreted by the sending cell, default is TRUE
#' @return A data frame of filtered interactions
#' @export
filter_lr_interactions = function(interactions = NULL, sender_celltypes = NULL, receiver_celltypes = NULL, secreted_lig = TRUE) {

      message("Filtering scSignalMap interactions")
      
      if (secreted_lig) {
          interactions_filtered = interactions %>%
                                  dplyr::filter(Ligand_secreted == TRUE,
                                  Receiver %in% receiver_celltypes,
                                  Sender %in% sender_celltypes,
                                  Sender != Receiver)
      } else {
          interactions_filtered = interactions %>%
                                  dplyr::filter(Receiver %in% receiver_celltypes,
                                  Sender %in% sender_celltypes,
                                  Sender != Receiver)
      }
      
      return(interactions_filtered)
}


#######################################################################
## Intersect Upregulated Receptors with Ligand-Receptor Interactions ##
#######################################################################

##' Intersect upregulated receptors with filtered ligand-receptor interactions
#'
#' This function takes previously identified DE receptor genes from chosen condition and identifies overlapping interactions
#'
#' @param upreg_receptors: name for output file containing DE receptors
#' @param interactions: ligand-receptor interactions dataframe from map_interactions 
#' @return A data frame of idendified DE receptor genes found in previously idendified interactions list
#' @export
intersect_upreg_receptors_with_lr_interactions = function(upreg_receptors = NULL, interactions = NULL) {
  
  message("Intersect upregulated receptors with filtered interactions")
  
  upreg_receptors_filtered_and_compared = upreg_receptors %>%
    dplyr::filter(gene_symbol %in% interactions$Receptor_Symbol)%>%
    dplyr::left_join(
      interactions %>% dplyr::select(Ligand_Symbol, Receptor_Symbol),
      by = c("gene_symbol" = "Receptor_Symbol"))
  
  return(upreg_receptors_filtered_and_compared)
}


##' Identify Enriched Pathways from DE Genes Using Enrichr
#'
#' This function performs pathway enrichment analysis on differentially expressed (DE) genes 
#' using the Enrichr. It filters enrichment results by an adjusted p-value cutoff and 
#' saves both individual and combined results for selected databases. The function uses a 
#' user-provided background gene set (from a Seurat object) and compares enrichment results 
#' to input DE genes to highlight relevant pathways.
#'
#' @param seurat_obj: seurat object containing genes for Enrichr to use as background, likely to be seurat object used as input for scSignalMap's MapInteractions function, must provide input
#' @param de_condition_filtered: DE genes output from find_markers_btwn_cond_for_celltype function, used to extract genes for Enrichr use
#' @param enrichr_databases: databases that Enrichr will use for pathway idenification, default includes c("BioCarta_2016", "GO_Biological_Process_2025", "KEGG_2021_Human", "NCI-Nature_2016", "WikiPathways_2024_Human")
#' @param adj_p_val_method: method for p-value adjustments, default is Benjamini-Hochberg (BH) 
#' @param adj_p_val_cutoff: desired cutoff for the adjusted p-value, default is 0.05
#' @return A data frame containing identified pathways, associated statistical values, common genes, etc.
#' @export
find_enriched_pathways = function(seurat_obj = NULL, de_condition_filtered = NULL, sender_celltypes= NULL, receiver_celltypes = NULL, enrichr_databases = c("BioCarta_2016", "GO_Biological_Process_2025", "KEGG_2021_Human", "NCI-Nature_2016", "WikiPathways_2024_Human"), adj_p_val_method = "BH", adj_p_val_cutoff = 0.05, ensdb = 'EnsDb.Hsapiens.v86') {
  
  # Get a list of genes for functional enrichment
  genes = unique(as.character(de_condition_filtered$gene_symbol))

  # Named list to convert gene names
  convert_me = genes
  names(convert_me) = toupper(genes)

  # Generate a background gene list based on all genes in Seurat object
  background_genes = rownames(seurat_obj[["RNA"]])
  background_genes = AnnotationDbi::select(eval(parse(text=ensdb)), keys = background_genes, keytype = "GENEID", columns = c("SYMBOL"))

  # Conduct enrichment analysis using Enrichr
  enrichment_results = enrichR::enrichr(genes, enrichr_databases, background = background_genes[,'SYMBOL'])

  # Add adjusted p-value
  for (db in names(enrichment_results)) {
    data = enrichment_results[[db]]
    data$Adjusted.P.value = p.adjust(data$P.value, method = adj_p_val_method)
    enrichment_results[[db]] = data
  }
  receiver_celltypes_clean = gsub("[.////]", "", receiver_celltypes)
  sender_celltypes_clean = gsub("[.////]", "", sender_celltypes)
  directory = paste0("enrichr_results/", sender_celltypes_clean, "_", receiver_celltypes_clean, "/")
  if (!dir.exists(directory)) {
    dir.create(directory)
  }
  for (db in names(enrichment_results)) {
    output = paste0(directory, sender_celltypes_clean, "_", receiver_celltypes_clean, "_", db, ".csv")
    write.csv(enrichment_results[[db]], file = output, row.names = FALSE)
  }
  
  enrichment_results_combined = bind_rows(
    lapply(names(enrichment_results), function(db) {
      df = enrichment_results[[db]]
      df$database = db
      df })
  )
  
  # Rename the Genes so they are correct for mouse, and shouldn't hurt human
  enrichr_results = enrichment_results_combined %>% mutate(tmp = paste(convert_me[strsplit(Genes, ";")[[1]]], collapse=';'))
  enrichr_results = enrichr_results[enrichr_results$Adjusted.P.value < adj_p_val_cutoff, ]

  return(enrichr_results)
}


#' Filter Enrichr results to keep only pathways containing upregulated receptors
#'
#' Takes the full Enrichr results and the upregulated receptors (after intersection with LR pairs)
#' and keeps only those pathways that contain at least one of the receptor gene symbols.
#' It adds a column "Matching_Receptors" with the matching receptor(s) and overwrites/saves
#' the filtered enrichr results.
#'
#' @param enrichr_results Dataframe returned by find_enriched_pathways (combined across databases)
#' @param upreg_receptors_filtered_and_compared Dataframe from intersect_upreg_receptors_with_lr_interactions
#' @param directory Directory where the filtered file should be saved
#' @param sender_clean Cleaned sender cell type name (for filename)
#' @param receiver_clean Cleaned receiver cell type name (for filename)
#' @param suffix Suffix for the output file, default "_enrichr_results_filtered.csv"
#' @return The filtered enrichr_results dataframe (also saved to disk)
#' @export
filter_enrichr_by_upreg_receptors = function(enrichr_results, upreg_receptors_filtered_and_compared) {

  # Safety checks
  if (is.null(enrichr_results) || nrow(enrichr_results) == 0) {
    message("No enrichr results to filter.")
    return(NULL)
  }
  if (is.null(upreg_receptors_filtered_and_compared) || nrow(upreg_receptors_filtered_and_compared) == 0) {
    message("No upregulated receptors – skipping enrichr filtering.")
    return(NULL)
  }

  # Standardise column names (in case Adjusted.P.value vs Adjusted_P_value)
  if ("Adjusted.P.value" %in% colnames(enrichr_results)) {
    enrichr_results = enrichr_results %>%
      dplyr::rename(Adjusted_P_value = Adjusted.P.value)
  }
  if ("Combined.Score" %in% colnames(enrichr_results)) {
    enrichr_results = enrichr_results %>%
      dplyr::rename(Combined_Score = Combined.Score)
  }

  # Unique receptor symbols (case-insensitive matching)
  receptor_symbols = tolower(unique(upreg_receptors_filtered_and_compared$gene_symbol))

  # Split Genes column into list and lowercase for matching
  enrichr_results = enrichr_results %>%
    dplyr::mutate(Genes_list = lapply(strsplit(Genes, ";"), function(x) trimws(tolower(x))))

  # Find matching receptors (original case)
  enrichr_results = enrichr_results %>%
    dplyr::mutate(Matching_Receptors = sapply(Genes_list, function(pathway_genes) {
      matches = intersect(pathway_genes, receptor_symbols)
      if (length(matches) == 0) return("")
      # Map back to original case
      orig = upreg_receptors_filtered_and_compared$gene_symbol[
        tolower(upreg_receptors_filtered_and_compared$gene_symbol) %in% matches
      ]
      paste(unique(orig), collapse = ";")
    })) %>%
    dplyr::select(-Genes_list)

  # Keep only pathways with at least one matching receptor
  enrichr_filtered = enrichr_results %>%
    dplyr::filter(Matching_Receptors != "" & !is.na(Matching_Receptors))
  message("Filtered to ", nrow(enrichr_filtered), " pathways containing upregulated receptors.")
  return(enrichr_filtered)
}


#' Export scSignalMap results in Neo4J-friendly format
#'
#' Takes the full MapInteractions output and creates three clean CSV files:
#' 1. senders_ligands: One row per sender-ligand with expression stats
#' 2. ligands_receptor_pairs: One row per valid L-R pair with stats
#' 3. receptors_receivers: One row per receptor-receiver with expression stats
#'
#' These are ideal for importing into Neo4j as nodes and relationships.
#'
#' @param interactions Dataframe from MapInteractions() (full pairs_data)
#' @param output_dir Directory to save the three CSV files (default: "Neo4J/")
#' @param prefix Prefix for filenames (e.g., "Q1_norm_test")
#' @return Invisibly returns list of file paths created
#' @export
export_for_neo4j = function(
  interactions,
  output_dir = "Neo4J/",
  prefix = "scSignalMap"
) {

  if (!dir.exists(output_dir)) dir.create(output_dir, recursive = TRUE)

  # 1. Sender - Ligand
  sender_ligands = interactions %>%
    dplyr::select(
      Sender,
      Ligand_Symbol,
      Ligand_Cells_Exp,
      Ligand_Avg_Exp,
      Ligand_secreted,
      Ligand_Counts,
      Ligand_gte_3,
      Ligand_gte_10
    ) #%>%
    #dplyr::distinct()

  sender_file = file.path(output_dir, paste0(prefix, "_senders_ligands.csv"))
  write.csv(sender_ligands, sender_file, row.names = FALSE)
  message("Saved senders to ligands: ", sender_file)

  # 2. Ligand - Receptor pairs
  lr_pairs = interactions %>%
    dplyr::select(
      Ligand_Symbol,
      Ligand_Counts,
      Ligand_gte_3,
      Ligand_gte_10,
      Receptor_Symbol,
      Receptor_Counts,
      Receptor_gte_3,
      Receptor_gte_10
    ) #%>%
    #dplyr::distinct()

  lr_file = file.path(output_dir, paste0(prefix, "_ligands_receptor_pairs.csv"))
  write.csv(lr_pairs, lr_file, row.names = FALSE)
  message("Saved L-R pairs: ", lr_file)

  # 3. Receptor - Receiver
  receiver_receptors = interactions %>%
    dplyr::select(
      Receptor_Symbol,
      Receptor_Counts,
      Receptor_gte_3,
      Receptor_gte_10,
      Receptor_Cluster_Marker,
      Receiver,
      Receptor_Avg_Exp
    ) #%>%
    #dplyr::distinct()

  receiver_file = file.path(output_dir, paste0(prefix, "_receptors_receivers.csv"))
  write.csv(receiver_receptors, receiver_file, row.names = FALSE)
  message("Saved receptors to receivers: ", receiver_file)

  invisible(list(
    sender_ligands = sender_file,
    ligand_receptor_pairs = lr_file,
    receiver_receptors = receiver_file
  ))
}

                                      
#' Generate Neo4j LOAD CSV script using local files from scSignalMap pipeline
#'
#' Automatically finds:
#' - The three core files in Neo4J/ (*_senders_ligands.csv, etc.)
#' - All filtered enrichr files in Neo4J/ (*_enrichr_results_DATABASE2.csv)
#' and generates a complete Cypher script using file:/// paths.
#'
#' @param neo4j_dir Directory with all CSV files (default: "Neo4J/")
#' @param dataset_name A string identifying this dataset (e.g., "Control", "Treatment_A").
#'        Required — will be added as a 'dataset' property on every node.
#' @param output_file Where to save the .cypher script (default: "Neo4J/load_scSignalMap.cypher")
#' @return Invisibly returns path to generated script
#' @export
generate_neo4j_local_load_script = function(
  neo4j_dir = "Neo4J/",
  dataset_name,
  output_file = file.path(neo4j_dir, "load_scSignalMap_local.cypher")
) {
  if (missing(dataset_name) || is.null(dataset_name) || dataset_name == "") {
    stop("dataset_name must be provided and non-empty.")
  }

  # Ensure directory ends with /
  neo4j_dir <- gsub("/+$", "/", neo4j_dir)

  # List all CSV files (relative names only)
  all_files <- list.files(neo4j_dir, pattern = "\\.csv$", full.names = FALSE)

  # Identify core files
  sender_file   <- all_files[grep("_senders_ligands\\.csv$", all_files)]
  lr_file       <- all_files[grep("_ligands_receptor_pairs\\.csv$", all_files)]
  receiver_file <- all_files[grep("_receptors_receivers\\.csv$", all_files)]
  pathway_files <- all_files[grep("_enrichr_results_DATABASE2\\.csv$", all_files)]

  if (length(sender_file) == 0 || length(lr_file) == 0 || length(receiver_file) == 0) {
    stop("Missing one or more core files in ", neo4j_dir, ". Found:\n", paste(all_files, collapse = "\n"))
  }

  cypher <- c(
    "// ================================================",
    "// Auto-generated Neo4j import script for scSignalMap (LOCAL VERSION)",
    paste0("// Dataset: ", dataset_name),
    "// Generated on: ", Sys.Date(),
    "// All files are in one folder - copy everything to Neo4j import/",
    "// ================================================\n",
    paste0(":param dataset_name => '", dataset_name, "';\n"),
    "// Indexes",
    "CREATE INDEX IF NOT EXISTS FOR (s:Sender) ON (s.name);",
    "CREATE INDEX IF NOT EXISTS FOR (l:Ligand_Symbol) ON (l.name);",
    "CREATE INDEX IF NOT EXISTS FOR (p:Receptor_Symbol) ON (p.name);",
    "CREATE INDEX IF NOT EXISTS FOR (c:Receiver) ON (c.name);",
    "CREATE INDEX IF NOT EXISTS FOR (z:Signal_Pathway_Info) ON (z.name);\n",

    "// 1. Sender → Ligand_Symbol",
    paste0('LOAD CSV WITH HEADERS FROM "file:///', sender_file, '" AS row'),
    "MERGE (s:Sender {name: row.Sender})",
    "  SET s.dataset = coalesce(s.dataset, $dataset_name)",
    "MERGE (l:Ligand_Symbol {name: row.Ligand_Symbol})",
    "  ON CREATE SET l += {",
    "    counts: toInteger(coalesce(row.Ligand_Counts, 0)),",
    "    l_exp_lvl_3: toFloat(coalesce(row.Ligand_gte_3, 0.0)),",
    "    l_exp_lvl_10: toFloat(coalesce(row.Ligand_gte_10, 0.0)),",
    "    Ligand_secreted: toBoolean(coalesce(row.Ligand_secreted, false)),",
    "    dataset: $dataset_name",
    "  }",
    "  SET l.dataset = coalesce(l.dataset, $dataset_name)",
    "MERGE (s)-[r:sender2ligand]->(l)",
    "  ON CREATE SET r.Ligand_Avg_Exp = toFloat(coalesce(row.Ligand_Avg_Exp, 0.0));\n",

    "// 2. Ligand_Symbol → Receptor_Symbol",
    paste0('LOAD CSV WITH HEADERS FROM "file:///', lr_file, '" AS row'),
    "MERGE (l:Ligand_Symbol {name: row.Ligand_Symbol})",
    "  SET l.dataset = coalesce(l.dataset, $dataset_name)",
    "MERGE (p:Receptor_Symbol {name: row.Receptor_Symbol})",
    "  ON CREATE SET p += {",
    "    counts: toInteger(coalesce(row.Receptor_Counts, 0)),",
    "    r_exp_lvl_3: toFloat(coalesce(row.Receptor_gte_3, 0.0)),",
    "    r_exp_lvl_10: toFloat(coalesce(row.Receptor_gte_10, 0.0)),",
    "    dataset: $dataset_name",
    "  }",
    "  SET p.dataset = coalesce(p.dataset, $dataset_name)",
    "MERGE (l)-[:ligand2receptorsymbol]->(p);\n",

    "// 3. Receptor_Symbol → Receiver",
    paste0('LOAD CSV WITH HEADERS FROM "file:///', receiver_file, '" AS row'),
    "MERGE (p:Receptor_Symbol {name: row.Receptor_Symbol})",
    "  ON CREATE SET p.Receptor_Cluster_Marker = toBoolean(coalesce(row.Receptor_Cluster_Marker, false))",
    "  SET p.dataset = coalesce(p.dataset, $dataset_name)",
    "MERGE (c:Receiver {name: row.Receiver})",
    "  SET c.dataset = coalesce(c.dataset, $dataset_name)",
    "MERGE (p)-[r:receptor2receivecluster]->(c)",
    "  ON CREATE SET r.Receptor_Avg_Exp = toFloat(coalesce(row.Receptor_Avg_Exp, 0.0));\n"
  )

  if (length(pathway_files) > 0) {
    unwinds <- character()
    for (f in pathway_files) {
      receiver <- sub(".*_", "", sub("_enrichr_results_DATABASE2\\.csv$", "", f))
      receiver <- gsub("[.////]", "", receiver)
      unwinds <- c(unwinds, paste0(' {file: "', f, '", receiver: "', receiver, '"}'))
    }

    cypher <- c(cypher,
      "// 4. Receiver → Signal_Pathway_Info (filtered pathways)",
      "UNWIND [",
      paste(unwinds, collapse = ",\n"),
      "] AS item",
      'LOAD CSV WITH HEADERS FROM "file:///" + item.file AS row',
      "WITH item.receiver AS recvName, row",
      "WHERE row.Term IS NOT NULL",
      "MERGE (c:Receiver {name: recvName})",
      "  SET c.dataset = coalesce(c.dataset, $dataset_name)",
      "MERGE (z:Signal_Pathway_Info {name: row.Term})",
      "  ON CREATE SET z += {",
      "    Database: coalesce(row.database, 'Unknown'),",
      "    Pathway_Adjusted_P_Value: toFloat(coalesce(row.Adjusted_P_value, 999)),",
      "    Combined_Score: toFloat(coalesce(row.Combined_Score, 0.0)),",
      "    Matching_Receptors: split(coalesce(row.Matching_Receptors, ''), ';'),",
      "    dataset: $dataset_name",
      "  }",
      "  SET z.dataset = coalesce(z.dataset, $dataset_name)",
      "MERGE (c)-[r:receiver2signalpathway]->(z)",
      "  ON CREATE SET r.Addl_Linked_Genes = split(coalesce(row.Genes, ''), ';');"
    )
  }

  # Save script
  dir.create(dirname(output_file), showWarnings = FALSE, recursive = TRUE)
  writeLines(cypher, output_file)

  message("Local Neo4j load script generated: ", output_file)
  message("Dataset tag: ", dataset_name)
  message("Next steps:")
  message("  1. Copy all CSV files from '", neo4j_dir, "' into your Neo4j 'import/' directory")
  message("  2. Run the generated script in Neo4j Browser")

  invisible(output_file)
}


#' Generate Neo4j LOAD CSV script for cloud instances (Sandbox/Aura) using HTTPS URLs
#'
#' This version is for Neo4j cloud deployments (Sandbox, AuraDB) where local file:/// paths are not supported.
#' Instead, provide direct public HTTPS URLs to your CSV files hosted on Google Drive (or GitHub, Dropbox, etc.).
#'
#' @param file_urls A named vector where names are the original filenames (as expected by the script)
#'                  and values are the direct HTTPS URLs (e.g., Google Drive download links).
#'                  Example: c("my_senders_ligands.csv" = "https://drive.google.com/uc?id=abc123&export=download", ...)
#' @param dataset_name A string identifying this dataset (e.g., "Control", "Treatment_A").
#'        Will be added as a 'dataset' property on all nodes.
#' @param output_file Path to save the generated .cypher script (default: "Neo4J/load_scSignalMap_cloud.cypher")
#' @return Invisibly returns the path to the generated script
#' @export
generate_neo4j_cloud_load_script = function(
  file_urls,
  dataset_name = "Default",
  output_file = "Neo4J/load_scSignalMap_cloud.cypher"
) {
  # Validate that required core files are provided
  required_core = c(
    grep("_senders_ligands\\.csv$", names(file_urls), value = TRUE),
    grep("_ligands_receptor_pairs\\.csv$", names(file_urls), value = TRUE),
    grep("_receptors_receivers\\.csv$", names(file_urls), value = TRUE)
  )
  if (length(required_core) < 3) {
    stop("Missing one or more core files in file_urls. Need: *_senders_ligands.csv, *_ligands_receptor_pairs.csv, *_receptors_receivers.csv")
  }
  
  sender_url = file_urls[grep("_senders_ligands\\.csv$", names(file_urls))]
  lr_url     = file_urls[grep("_ligands_receptor_pairs\\.csv$", names(file_urls))]
  receiver_url = file_urls[grep("_receptors_receivers\\.csv$", names(file_urls))]
  
  # Pathway files (optional)
  pathway_urls = file_urls[grep("_enrichr_results_DATABASE2\\.csv$", names(file_urls))]
  
  cypher = c(
    "// ================================================",
    "// Auto-generated Neo4j import script for scSignalMap (CLOUD VERSION)",
    paste0("// Dataset: ", dataset_name),
    "// Generated on: ", Sys.Date(),
    "// Uses direct HTTPS URLs (e.g., Google Drive direct download links)",
    "// Run this directly in Neo4j Browser on your Sandbox/Aura instance",
    "// ================================================\n",
    ":param dataset_name => $dataset_name;\n",
    "// Indexes",
    "CREATE INDEX IF NOT EXISTS FOR (s:Sender) ON (s.name);",
    "CREATE INDEX IF NOT EXISTS FOR (l:Ligand_Symbol) ON (l.name);",
    "CREATE INDEX IF NOT EXISTS FOR (p:Receptor_Symbol) ON (p.name);",
    "CREATE INDEX IF NOT EXISTS FOR (c:Receiver) ON (c.name);",
    "CREATE INDEX IF NOT EXISTS FOR (z:Signal_Pathway_Info) ON (z.name);\n",
    
    "// 1. Sender to Ligand_Symbol",
    paste0('LOAD CSV WITH HEADERS FROM "', sender_url, '" AS row'),
    "MERGE (s:Sender {name: row.Sender})",
    " ON CREATE SET s.dataset = $dataset_name",
    " SET s.dataset = coalesce(s.dataset, $dataset_name)",
    "MERGE (l:Ligand_Symbol {name: row.Ligand_Symbol})",
    " ON CREATE SET l += {",
    " counts: toInteger(coalesce(row.Ligand_Counts, 0)),",
    " l_exp_lvl_3: toFloat(coalesce(row.Ligand_gte_3, 0.0)),",
    " l_exp_lvl_10: toFloat(coalesce(row.Ligand_gte_10, 0.0)),",
    " Ligand_secreted: toBoolean(coalesce(row.Ligand_secreted, false))",
    " dataset: $dataset_name",
    " }",
    " SET l.dataset = coalesce(l.dataset, $dataset_name)",
    "MERGE (s)-[r:sender2ligand]->(l)",
    " ON CREATE SET r.Ligand_Avg_Exp = toFloat(coalesce(row.Ligand_Avg_Exp, 0.0));\n",
    
    "// 2. Ligand_Symbol to Receptor_Symbol",
    paste0('LOAD CSV WITH HEADERS FROM "', lr_url, '" AS row'),
    "MERGE (l:Ligand_Symbol {name: row.Ligand_Symbol})",
    " SET l.dataset = coalesce(l.dataset, $dataset_name)",
    "MERGE (p:Receptor_Symbol {name: row.Receptor_Symbol})",
    " ON CREATE SET p += {",
    "  counts: toInteger(coalesce(row.Receptor_Counts, 0)),",
    "  r_exp_lvl_3: toFloat(coalesce(row.Receptor_gte_3, 0.0)),",
    "  r_exp_lvl_10: toFloat(coalesce(row.Receptor_gte_10, 0.0))",
    "  dataset: $dataset_name",
    " }",
    " SET p.dataset = coalesce(p.dataset, $dataset_name)",
    "MERGE (l)-[:ligand2receptorsymbol]->(p);\n",
    
    "// 3. Receptor_Symbol to Receiver",
    paste0('LOAD CSV WITH HEADERS FROM "', receiver_url, '" AS row'),
    "MERGE (p:Receptor_Symbol {name: row.Receptor_Symbol})",
    "  ON CREATE SET p.Receptor_Cluster_Marker = toBoolean(coalesce(row.Receptor_Cluster_Marker, false))",
    "  SET p.dataset = coalesce(p.dataset, $dataset_name)",
    "MERGE (c:Receiver {name: row.Receiver})",
    "  SET c.dataset = coalesce(c.dataset, $dataset_name)",
    "MERGE (p)-[r:receptor2receivecluster]->(c)",
    "  ON CREATE SET r.Receptor_Avg_Exp = toFloat(coalesce(row.Receptor_Avg_Exp, 0.0));\n"
  )
  
  if (length(pathway_urls) > 0) {
    unwinds = character()
    for (i in seq_along(pathway_urls)) {
      f_name = names(pathway_urls)[i]
      url = pathway_urls[i]
      # Extract receiver from filename (same logic as original)
      receiver = sub(".*_", "", sub("_enrichr_results_DATABASE2\\.csv$", "", f_name))
      receiver = gsub("[.////]", "", receiver)
      unwinds = c(unwinds, paste0(' {url: "', url, '", receiver: "', receiver, '"}'))
    }
    cypher = c(cypher,
      "// 4. Receiver to Signal_Pathway_Info (filtered pathways)",
      "UNWIND [",
      paste(unwinds, collapse = ",\n"),
      "] AS item",
      'LOAD CSV WITH HEADERS FROM item.url AS row',
      "WITH item.receiver AS recvName, row",
      "WHERE row.Term IS NOT NULL",
      "MERGE (c:Receiver {name: recvName})",
      "  SET c.dataset = coalesce(c.dataset, $dataset_name)",
      "MERGE (z:Signal_Pathway_Info {name: row.Term})",
      "  ON CREATE SET z += {",
      "    Database: coalesce(row.database, 'Unknown'),",
      "    Pathway_Adjusted_P_Value: toFloat(coalesce(row.Adjusted_P_value, 999)),",
      "    Combined_Score: toFloat(coalesce(row.Combined_Score, 0.0)),",
      "    Matching_Receptors: split(coalesce(row.Matching_Receptors, ''), ';'),",
      "    dataset: $dataset_name",
      "  }",
      "  SET z.dataset = coalesce(z.dataset, $dataset_name)",
      "MERGE (c)-[r:receiver2signalpathway]->(z)",
      "  ON CREATE SET r.Addl_Linked_Genes = split(coalesce(row.Genes, ''), ';');"
    )
  }
  
  dir.create(dirname(output_file), showWarnings = FALSE, recursive = TRUE)
  writeLines(cypher, output_file)
  message("Cloud Neo4j load script generated: ", output_file)
  message("Run this script directly in your Neo4j Sandbox/Aura Browser.")
  invisible(output_file)
}
                                      
#' Run Full scSignalMap Pipeline 
#'
#' This function performs a full ligand–receptor signaling analysis  
#' using scSignalMap and Seurat. It performs ligand–receptor interaction mapping, 
#' differential expression analysis, receptor filtering, receptor–ligand 
#' intersection, and pathway enrichment analysis in one workflow.  
#'
#' @param seurat_obj: seurat object containing scRNA-seq data
#' @param prep_SCT: logical; whether to run PrepSCTFindMarkers before differential expression, default = TRUE
#' @param cond_column: meta.data column in Seurat object containing condition labels
#' @param cond_name1: condition column value that compared to cond_name2 for differential expression analysis, must be specified
#' @param cond_name2: condition column value that compared to cond_name1 for differential expression analysis, must be specified
#' @param celltype_column: meta.data column in Seurat object for celltype information
#' @param celltype_name: celltype column value of interest, must be specified
#' @param sender_celltypes: celltype of cell(s) that are "senders" in the cell-to-cell communication interaction, ex. "<celltype>" or c("celltype1", "celltype2")
#' @param receiver_celltypes: celltype of cell(s) that are "receivers" in the cell-to-cell communication interaction, ex. "<celltype>" or c("celltype1", "celltype2")
#' @param secreted_lig: logical; whether to restrict to secreted ligands in filtering, default = TRUE
#' @param FC_cutoff: log2 fold change cutoff for differential expression and receptor filtering, default = 0.3
#' @param adj_p_val_cutoff: adjusted p-value cutoff for differential expression and pathway enrichment, default = 0.05
#' @param enrichr_databases: databases that Enrichr will use for pathway idenification, default includes c("BioCarta_2016", "GO_Biological_Process_2025", "KEGG_2021_Human", "NCI-Nature_2016", "WikiPathways_2024_Human")
#' @param adj_p_val_method: method for p-value adjustments, default = "BH" 
#' @return a list containing ligand–receptor interactions, DE genes, upregulated receptors, filtered interactions, intersected receptors, and pathway enrichment results.
#' @export
run_full_scSignalMap_pipeline = function(
    seurat_obj = NULL, 
    prep_SCT = TRUE, 
    cond_column = NULL, 
    cond_name1 = NULL, 
    cond_name2 = NULL, 
    celltype_column = NULL, 
    celltype_name = NULL, 
    sender_celltypes = NULL, 
    receiver_celltypes = NULL, 
    secreted_lig = TRUE, 
    FC_cutoff = 0.3, 
    adj_p_val_cutoff = 0.05, 
    enrichr_databases = c("BioCarta_2016", "GO_Biological_Process_2025", "KEGG_2021_Human",
                          "NCI-Nature_2016", "WikiPathways_2024_Human"), 
    adj_p_val_method = "BH", 
    ensdb = 'EnsDb.Hsapiens.v86', 
    species='human'){
  
  #####################
  ### Run pipeline  ###
  #####################
  message("Running MapInteractions...")
  LR_interactions = MapInteractions(seurat_obj, 
                                    group_by = celltype_column,
                                    species=species)
  # Container for all pairwise results
  all_results = list()
  
  # progress bar for loops  
  total_pairs = length(sender_celltypes) * length(receiver_celltypes)
  pb = progress_bar$new(
    format = "[:bar] :current/:total (:percent) | :sender to :receiver | ETA: :eta\n",
    total = total_pairs,
    width = 80
  )
  pb$tick(0)  # initialize
  for (sender in sender_celltypes) {
    for(receiver in receiver_celltypes) {
      
      receiver_clean = gsub("[.////]", "", receiver)
      sender_clean = gsub("[.////]", "", sender)
      
      pb$tick(tokens = list(sender = sender_clean, receiver = receiver_clean))
      
      message("Processing ", sender_clean, " → ", receiver_clean, "...")
      
      message("Finding DE genes...")
      de_cond_celltype = find_markers_btwn_cond_for_celltype(
        seurat_obj = seurat_obj,
        prep_SCT = prep_SCT,
        cond_column = cond_column,
        cond_name1 = cond_name1,
        cond_name2 = cond_name2,
        celltype_column = celltype_column,
        celltype_name = receiver, # receiver is the one with DE
        FC_cutoff = FC_cutoff,
        adj_p_val_cutoff = adj_p_val_cutoff,
        ensdb = ensdb)
      
      message("Finding upregulated receptors...")
      upreg_receptors = find_upreg_receptors(
        de_condition_filtered = de_cond_celltype,
        FC_cutoff = FC_cutoff, 
        species=species)
      
      message("Filtering LR interactions...")
      interactions_filtered = filter_lr_interactions(
        interactions = LR_interactions,
        sender_celltypes = sender,
        receiver_celltypes = receiver,
        secreted_lig = secreted_lig)
      
      message("Intersecting receptors with interactions...")
      upreg_receptors_filtered_and_compared =
        intersect_upreg_receptors_with_lr_interactions(
          upreg_receptors = upreg_receptors,
          interactions = interactions_filtered)
      
      message("Running pathway enrichment...")
      enrichr_results = find_enriched_pathways(
        seurat_obj = seurat_obj,
        de_condition_filtered = de_cond_celltype,
        sender_celltypes = sender,
        receiver_celltypes = receiver,
        enrichr_databases = enrichr_databases,
        adj_p_val_method = adj_p_val_method,
        adj_p_val_cutoff = adj_p_val_cutoff,
        ensdb = ensdb)
      
      # Filter enrichr results to pathways involving upregulated receptors for Neo4J CSV
      enrichr_filtered = filter_enrichr_by_upreg_receptors(
        enrichr_results = enrichr_results,
        upreg_receptors_filtered_and_compared = upreg_receptors_filtered_and_compared
      )
     
      
      
      ###################################
      ### Return all results together ###
      ###################################
      pair_name = paste0(sender_clean, "_", receiver_clean)
      all_results[[pair_name]] = list(
        LR_interactions = LR_interactions,
        de_cond_celltype = de_cond_celltype,
        upreg_receptors = upreg_receptors,
        interactions_filtered = interactions_filtered,
        upreg_receptors_filtered_and_compared = upreg_receptors_filtered_and_compared,
        enrichr_results = enrichr_results,
        enrichr_filtered = enrichr_filtered,
        sender = sender,
        receiver = receiver,
        sender_clean = sender_clean,
        receiver_clean = receiver_clean)
    }
  }
  
  
  return(all_results)
}

#' Run Post-Processing & Neo4J Export
#'
#' Takes the output from run_scSignalMap_core() and saves all CSV files,
#' exports core Neo4J tables, and generates load scripts.
#'
#' @param all_results List returned by run_full_scSignalMap_pipeline()
#' @param neo4j_prefix Prefix for core Neo4J CSV filenames (default: "scSignalMap")
#' @param upload_to_drive Logical; upload Neo4J CSVs to Google Drive and generate cloud script
#' @param drive_folder_name Name of Google Drive folder (default: auto-generated with date)
#' @param generate_local_script Logical; create local file:/// load script
#' @param generate_cloud_script Logical; create HTTPS cloud load script (auto-enabled if upload_to_drive=TRUE)
#' @export
run_post_processing_Neo4J = function(
    all_results,
    neo4j_prefix = "scSignalMap",
    dataset_name = "default",
    upload_to_drive = FALSE,
    drive_folder_name = NULL,
    generate_local_script = TRUE,
    generate_cloud_script = NULL
) {
  if (is.null(generate_cloud_script)) {
    generate_cloud_script = upload_to_drive
  }
  
  # Ensure directories exist
  dir.create("enrichr_results", showWarnings = FALSE, recursive = TRUE)
  dir.create("Neo4J", showWarnings = FALSE, recursive = TRUE)
  
  # Grab the global LR_interactions (same for all pairs)
  LR_interactions = all_results[[1]]$LR_interactions
  
  message("Saving per-pair CSV files and filtered enrichr results...")
  
  for (pair_name in names(all_results)) {
    res = all_results[[pair_name]]
    sender_clean   = res$sender_clean
    receiver_clean = res$receiver_clean
    
    dir_pair = file.path("enrichr_results", paste0(sender_clean, "_", receiver_clean))
    dir.create(dir_pair, showWarnings = FALSE, recursive = TRUE)
    
    # Save individual results
    write.csv(res$LR_interactions, file.path(dir_pair, paste0(sender_clean, "_", receiver_clean, "_LR_interactions2.csv")), row.names = FALSE)
    write.csv(res$de_cond_celltype, file.path(dir_pair, paste0(sender_clean, "_", receiver_clean, "_de_cond_celltype2.csv")), row.names = FALSE)
    write.csv(res$upreg_receptors, file.path(dir_pair, paste0(sender_clean, "_", receiver_clean, "_upreg_receptors2.csv")), row.names = FALSE)
    write.csv(res$interactions_filtered, file.path(dir_pair, paste0(sender_clean, "_", receiver_clean, "_interactions_filtered2.csv")), row.names = FALSE)
    write.csv(res$upreg_receptors_filtered_and_compared, file.path(dir_pair, paste0(sender_clean, "_", receiver_clean, "_upreg_receptors_filtered_and_compared2.csv")), row.names = FALSE)
    write.csv(res$enrichr_results, file.path(dir_pair, paste0(sender_clean, "_", receiver_clean, "_enrichr_results2.csv")), row.names = FALSE)
    
    # Save filtered enrichr for Neo4J (only if non-empty)
    if (!is.null(res$enrichr_filtered) && nrow(res$enrichr_filtered) > 0) {
      neo4j_file = file.path("Neo4J",
                             paste0(sender_clean, "_", receiver_clean, "_enrichr_results_DATABASE2.csv"))
      write.csv(res$enrichr_filtered, neo4j_file, row.names = FALSE)
      message("Saved ", nrow(res$enrichr_filtered),
              " signaling-relevant pathways for ", pair_name)
    } else {
      message("No signaling-relevant pathways for ", pair_name, " (skipped)")
    }
  }
  
  # Export the three core Neo4J tables (from full LR_interactions)
  export_for_neo4j(
    interactions = LR_interactions,
    output_dir = "Neo4J/",
    prefix = neo4j_prefix
  )
  
  # Generate scripts
  if (generate_local_script) {
    generate_neo4j_local_load_script(
      neo4j_dir = "Neo4J/",
      dataset_name,
      output_file = "Neo4J/load_scSignalMap_local.cypher"
    )
  }
  
  if (upload_to_drive) {
    if (!requireNamespace("googledrive", quietly = TRUE)) {
      stop("Package 'googledrive' needed for upload_to_drive = TRUE.")
    }
    library(googledrive)
    
    message("\n=== Uploading Neo4J files to Google Drive ===")
    drive_auth()
    
    if (is.null(drive_folder_name)) {
      drive_folder_name = paste0("scSignalMap_Neo4j_", format(Sys.Date(), "%Y%m%d"))
    }
    
    drive_folder = try(drive_get(drive_folder_name), silent = TRUE)
    if (inherits(drive_folder, "try-error") || nrow(drive_folder) == 0) {
      drive_folder = drive_mkdir(drive_folder_name)
      message("Created folder: ", drive_folder_name)
    } else {
      drive_folder = drive_folder[1, ]
    }
    
    local_files = list.files("Neo4J/", pattern = "\\.csv$", full.names = TRUE)
    if (length(local_files) == 0) {
      warning("No CSV files in Neo4J/ to upload.")
    } else {
      message("Uploading ", length(local_files), " files...")
      uploaded = lapply(local_files, function(f) drive_upload(media = f, path = drive_folder, overwrite = TRUE))
      uploaded = do.call(rbind, uploaded)
      
      drive_share(uploaded, role = "reader", type = "anyone")
      
      file_urls = setNames(
        paste0("https://drive.google.com/uc?id=", uploaded$id, "&export=download"),
        uploaded$name
      )
      
      message("Direct download URLs:")
      print(file_urls)
      
      if (generate_cloud_script) {
        generate_neo4j_cloud_load_script(
          file_urls = file_urls,
          dataset_name,
          output_file = "Neo4J/load_scSignalMap_cloud.cypher"
        )
        message("Cloud load script generated: Neo4J/load_scSignalMap_cloud.cypher")
      }
    }
  }
  
  message("Post-processing and Neo4J export complete.")
  invisible(all_results)
}                                      
##' Create Master Interaction List
#'
#' This function creates master interaction list by combining DE ligands/receptors, Enrichr results, and scSignalMap interactions
#'
#' @param enrichr_results Data frame of Enrichr pathway enrichment results, default is results$enrichr_results 
#' @param de_receptors Data frame of upregulated receptors, default is results$upreg_receptors_filtered_and_compared
#' @param scSignalMap_data_filtered Data frame of filtered interactions, default is results$interactions_filtered
#' @return A data frame containing merged ligand/receptor info, enrichment, and interaction data
#' @export
create_master_interaction_list = function(
  enrichr_results = results$enrichr_results,
  de_receptors = results$upreg_receptors_filtered_and_compared,
  scSignalMap_data_filtered = results$interactions_filtered
) {
  
  ## Step 1: Clean Enrichr results
  enrichr_results = enrichr_results[, !(names(enrichr_results) %in% c("Old.P.value", "Old.Adjusted.P.value"))]
  
  # Expand genes column
  expanded_enrichr = enrichr_results %>%
    tidyr::separate_rows(Genes, sep = ";") %>%
    dplyr::mutate(Genes = stringr::str_trim(Genes))
  
  # Find common genes
  common_genes = intersect(unique(expanded_enrichr$Genes), unique(de_receptors$gene_symbol))
  
  # Filter Enrichr results to only include common genes
  enrichr_results = enrichr_results %>%
    dplyr::filter(sapply(Genes, function(x) {
      any(stringr::str_trim(unlist(strsplit(x, ";"))) %in% common_genes)
    }))
  
  # Filter DE receptors to only include common genes
  de_receptors = de_receptors %>%
    dplyr::filter(gene_symbol %in% common_genes)
  
  ## Step 2: Build master list with DE receptor info
  master_list = de_receptors[, c("gene_symbol", "avg_log2FC", "p_val_adj")]
  colnames(master_list)[colnames(master_list) == "gene_symbol"] = "Receptor_Symbol"
  
  # Prepare Enrichr info for merging
  expanded_enrichr_subset = expanded_enrichr %>%
    dplyr::select(Genes, Term, Adjusted.P.value) %>%
    dplyr::rename(Receptor_Symbol = Genes, enrichr_p_val_adj = Adjusted.P.value)
  
  # Merge DE receptor info with Enrichr results
  master_list = dplyr::left_join(master_list, expanded_enrichr_subset, by = "Receptor_Symbol")
  
  ## Step 3: Merge with scSignalMap interactions
  master_list = merge(
    master_list,
    scSignalMap_data_filtered,
    by.x = "Receptor_Symbol",
    by.y = "Receptor_Symbol",
    all.x = TRUE,
    sort = FALSE
  )
  
  # Clean up: remove rows without receptor info
  master_list = master_list[!is.na(master_list$Receptor_Symbol), ]
  
  # Remove unwanted column "X" if it exists
  if ("X" %in% colnames(master_list)) master_list$X = NULL
  
  # Reorder columns so receptor symbol/info come first
  if ("Receptor" %in% colnames(master_list)) {
    cols = colnames(master_list)
    new_order = c("Receptor_Symbol", "Receptor", setdiff(cols, c("Receptor_Symbol", "Receptor")))
    master_list = master_list[, new_order]
  }
  
  return(master_list)
}


